File: hackathon/security-copilot/backend/app/models/source.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Source(Base):
    __tablename__ = "sources"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    url: Mapped[str] = mapped_column(String(2048), nullable=False)
    source_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    polling_interval_minutes: Mapped[int] = mapped_column(Integer, default=60, nullable=False)
    last_polled_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    # Relationships
    articles: Mapped[list["Article"]] = relationship(  # noqa: F821
        "Article", back_populates="source"
    )

File: hackathon/security-copilot/frontend/src/types/source.ts
export type SourceType = 'rss_feed' | 'news_site' | 'vendor_blog' | 'threat_feed';

export interface Source {
  id: number;
  name: string;
  url: string;
  source_type: SourceType;
  is_active: boolean;
  polling_interval_minutes: number;
  last_polled_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface SourceCreate {
  name: string;
  url: string;
  source_type: SourceType;
  is_active?: boolean;
  polling_interval_minutes?: number;
}

export interface SourceUpdate {
  name?: string;
  url?: string;
  source_type?: SourceType;
  is_active?: boolean;
  polling_interval_minutes?: number;
}

export interface CollectResult {
  message: string;
  collected: number;
  errors: number;
}
